### DISCUSSION TITLE
Lots of deprecations in Symfony 5.4 beta

### DISCUSSION THREAD

**Discussion by guillaumesmo** [symfony/symfony#44082] [Q&A]
While trying Symfony 5.4 beta in my application I get a lot of deprecations of the following type:
```
Method "*****" might add "***" as a native return type declaration in the future. Do the same in implementation "*****" now to avoid errors or add an explicit @return annotation to suppress this message
```

although I understand the message, I don't get why this deprecation is thrown for classes that don't implement Symfony classes/interfaces, and some are not even part of my codebase. Any idea if that is expected?

eg:
```
1x: Method "MyClass::getId()" might add "int" as a native return type declaration in the future. Do the same in child class "MyAbstractClass" now to avoid errors or add an explicit @return annotation to suppress this message.

```

seems to be related to #43021

**Comment by wouterj** [comment]
This is expected. The deprecation is triggered whenever a child class is missing a return type declaration. You must add the return type to the child before the parent method can add the return type (otherwise you get a fatal error).

See https://symfony.com/blog/preparing-your-apps-and-bundles-for-symfony-6 and https://wouterj.nl/2021/09/symfony-6-native-typing for more information about this. There is a little type patch script to add most of these return types automatically.

**Reply by stof** [reply]
@wouterj I think this report is related to the fact that DebugClassLoader currently reports that *any* class that defines a type in phpdoc is asking for its child classes to add the native return type, instead of being an opt-in. So here, it also requires child classes of `MyAbstractClass` to add a return type.

To avoid that, we would probably need a way for classes/interfaces to declare a tentative return type when they intend to change the type in the future (the phpdoc `@return` might still be used as the source of the type, but something else would need to opt-in for it to be a tentative return type).
The patch script is still an help, as you could migrate all your project classes to native types with it, but it still creates noise.

**Reply by wouterj** [reply]
It being opt-out (by adding `@return` to your own methods causing this notice - which the type patcher can also do for you) is a feature. Without it, every project other than Symfony is unable to add strict typing (or at least, do so with a smooth upgrade path).

So everyone can do 2 things:
1. Add the return type declaration to the method. This is future proof and allows the vendor library providing the parent class to use return types in the future.
2. Accept the risk that any return type added by a vendor can break your code by adding `@return` PHPdoc to each method (especially useful to opt-out from PHP types not yet supported on the version you're using).

Both can be done by the type patcher tool for the 90% common cases. During the past months, I've been successfully doing this on quite some open source projects (e.g. Laravel, Guzzle, Carbon, iter).

Note that a deprecation is not triggered when both child and parent class are maintained by the current application. In this situation, there is no risk as you're in full control of return type declarations for these methods. so `MyAbstractClass` and `MyClass` should not trigger a deprecation, but `TheirAbstractClass` and `MyClass` should.

**Reply by guillaumesmo** [reply]
My point was exactly that "a deprecation is not triggered when both child and parent class are maintained by the current application" doesn't seem to work

A deprecation is triggered for 2 classes that are in my sources. Also strange is that they appear in the "Other deprecation notices" section.

**Reply by wouterj** [reply]
Hmm, that is worth a look. Sorry for misunderstanding your question!

Do you maybe have time to debug a little bit what is going on? I think these are the related lines:

https://github.com/symfony/symfony/blob/2b31f0bf53565bbd0a04a20e518cb26a9f360822/src/Symfony/Component/ErrorHandler/DebugClassLoader.php#L358-L367

In this first block, `$class` is the "MyAbstractClass" if I'm correct.

https://github.com/symfony/symfony/blob/2b31f0bf53565bbd0a04a20e518cb26a9f360822/src/Symfony/Component/ErrorHandler/DebugClassLoader.php#L559

In the second code block, `strncmp($ns, $declaringClass, $len)` is the one checking whether this is a class in the same application. `$ns`/ `$len` are `$vendor`/`$vendorLen` from the first block and `$declaringClass` is the "MyClass" in your example.

**Reply by guillaumesmo** [reply]
thanks for the relevant excerpt, I think I understand where it comes from. 

In this case I'm working on a legacy project which still has classes without namespace (PEAR-style non-namespaced convention, as supported by composer https://getcomposer.org/doc/04-schema.md#psr-0)
when a class in my App namespace extends one of those classes or the other way around (`$vendor = ''` vs `$vendor = 'App\\'`) the deprecation is triggered for those cases

I can maybe suggest to disable deprecations when vendor is empty, but then you'll also silence deprecations when extending native interfaces like \ArrayAccess
or maybe only silence when the extending class' vendor is empty?
or just make all this opt-in/out somehow?

**Reply by stof** [reply]
the code linked here seems to handle the case of PEAR-style namespacing, by looking for the first `_` in the class name when there is no `\`.

Is your code actually using a PEAR-style naming, or just global class names without a PEAR-style namespace prefix ?

**Reply by guillaumesmo** [reply]
it indeed handles the PEAR-style namespacing (eg, for a class name `AAA_BBB_CCC`, `$vendor = 'AAA\'`), except that doesn't solve the issue of `AAA_BBB_CCC` not being in the same namespace as `ZZZ_BBB_CCC` although both are in my codebase

same if true for my `AAA_` vs `App\` classes which don't have the same namespace either

**Reply by stof** [reply]
Well, if your own code has both `AAA_BBB_CCC` and `ZZZ_BBB_CCC`, it means you don't use pear-style namespacing for your own code, as you don't put your code under a namespace prefix.

**Reply by guillaumesmo** [reply]
So I understand that any project with multiple namespace roots (with classes extending another namespace) will have to patch their files to silence deprecations in Symfony 5.4

At least it's clear now in which specific cases I will have to add return types, thanks for your help.

**Reply by gharlan** [reply]
@wouterj 

> Note that a deprecation is not triggered when both child and parent class are maintained by the current application.

What if both classes are from vendor libs? I get this message:

> Method "TCPDF::_getxobjectdict()" might add "string" as a native return type declaration in the future. Do the same in child class "setasign\Fpdi\Tcpdf\Fpdi" now to avoid errors or add an explicit `@return` annotation to suppress this message.

I don't feel comfortable to ask the Fpdi developer to add the return type. I think there is nothing wrong [in Fpdi](https://github.com/Setasign/FPDI/blob/273ab1544a6a53aa7beb6f14366586a9dd3369c5/src/Tcpdf/Fpdi.php#L141-L144).

**Reply by wouterj** [reply]
There is, there is no `@return` annotation on the method (inheritdocs is meant not to work here).

I do understand what you mean though. I also was very unsure about submitting PRs like that. But from my experience, these PRs were smooth sailing all the way: https://github.com/nikic/iter/pull/90, https://github.com/guzzle/psr7/pull/444, https://github.com/laravel/framework/pull/38819, https://github.com/briannesbitt/Carbon/pull/2453 (as you can see, even some packages that don't have any Symfony dependency).

At last, these deprecations should be marked as "indirect". These are much less important than "direct" deprecations.

**Comment by ahundiak** [comment]
Bit off-topic but I'd just like to repeat my plea for a simple and documented way to turn off deprecation notices for 5.4.  I understand the need for the notices but the sheer volume that will be generated is really going to freak out many developers.

**Reply by guillaumesmo** [reply]
or at least an easy way to turn this type of deprecations off in a centralized way, "It being opt-out (by adding @return to your own methods causing this notice ...)" is not going to satisfy those which will need to review my huge PR 😁
